src/contract-read.ts

```diff
@@ -28,6 +34,14 @@ export async function readContract(
     height = networkInfo.height;
   }
 
+  if (contractId in cache) {
+    if (height in cache[contractId]) {
+      const res = JSON.parse(cache[contractId][height]);
+
+      return returnValidity ? { state: res.state, validity: res.validity } : res.state;
```

Comment:
perhaps this would require unrelated changes. But returning both object of state and validity, or just state, creates an annoying typing situation. Could it be a return of `{ state: IState, validity?: IValidity }`, then validity can either be undefined or defined.